### PR TITLE
Merge Confirmation Dialogs into Alert, Confirmation or Prompt

### LINKED ISSUE #4103: Add a generic alert dialog using built in solution instead of browser alert
**Describe the solution you'd like**

Now that #4053 is merged, this arises the inconsistency of the `alert()` with the built in dialog prompts etc.

**Additional**

I'm leaving this here in case someone wants to take this on beforehand, but I am planning to add a new generic alert dialog this weekend (or before) if I get chance.

Very similar to the above mentioned PR where the `ha-paper-dialog` is invoked, but as an alert dialog instead of a prompt

### PR DESCRIPTION
**Pull request by timmo001**
## Description

Adds alert dialog and replaces the confirmation dialog so that prompts and alerts can be called in the same component. Also adds prompts to more sections of the UI.

## Issue this references

Closes #4103

## Screenshots

Dev tools - States (no entity_id):

image

Dev tools - Events (no event type)

image

Automation - Delete:

image

### CODE REVIEW

**Comment by timmo001** [comment]
There's a good few, but I'll leave it at that for now. :smiley:

File: src/dialogs/alert/dialog-alert.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    return html`
+      <ha-paper-dialog
+        with-backdrop
+        opened
+        @opened-changed="${this._openedChanged}"
+      >
+        <h2>
+          ${this._params.title && this._params.title}
```
Comment (bramkragten) [inline src/dialogs/alert/dialog-alert.ts]:
```suggestion
          ${this._params && this._params.title}
```

**Comment by bramkragten** [comment]
I wonder if we can't combine the dialogs, the only real difference is the cancel button?

**Comment by bramkragten** [comment]
And later also add a `prompt`

**Comment by timmo001** [comment]
I guess so. I'll take the android naming of an `AlertDialog` since that allows 1+ options

**Comment by timmo001** [comment]
Working correctly now:

image

No cancel or title on secondary dialog:
image

**Comment by bramkragten** [comment]
With prompt I meant this: https://developer.mozilla.org/en-US/docs/Web/API/Window/prompt

**Comment by bramkragten** [comment]
So we would have: `alert`, `confirm` and `prompt`

**Comment by bramkragten** [comment]
And I suggest we make 3 different helpers for the same dialog:
`showAlert`, `showConfirm` and `showPrompt`

File: src/dialogs/generic/show-dialog-generic.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  confirmation: boolean;
+  prompt: boolean;
+}
+
+export const loadGenericDialog = () =>
+  import(/* webpackChunkName: "confirmation" */ "./dialog-generic");
+
+export const showAlertDialog = (
+  element: HTMLElement,
+  systemLogDetailParams: AlertDialogParams
```
Comment (bramkragten) [inline src/dialogs/generic/show-dialog-generic.ts]: `systemLogDetailParams`? I think that is a left over from copy paste?

File: src/dialogs/generic/dialog-generic.ts [hunk 3]
```diff
@@ -14,16 +14,16 @@ import "../../components/dialog/ha-paper-dialog";
...
 import { HomeAssistant } from "../../types";
-import { ConfirmationDialogParams } from "./show-dialog-confirmation";
+import { GenericDialogParams } from "./show-dialog-generic";
 import { PolymerChangedEvent } from "../../polymer-types";
 import { haStyleDialog } from "../../resources/styles";
 
-@customElement("dialog-confirmation")
-class DialogConfirmation extends LitElement {
+@customElement("dialog-generic")
+class DialogGeneric extends LitElement {
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-generic.ts]: Let's call it `DialogBox`

File: src/dialogs/generic/show-dialog-generic.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  confirmText?: string;
+  cancel?: () => void;
+}
+
+export interface PromptDialogParams extends ConfirmationDialogParams {
+  inputType?: string;
+  inputHint?: string;
+}
+
+export interface GenericDialogParams extends ConfirmationDialogParams {
```
Comment (bramkragten) [inline src/dialogs/generic/show-dialog-generic.ts]: Shouldn't this be:
```ts
type GenericDialogParams = AlertDialogParams | ConfirmationDialogParams | PromptDialogParams;
```

**Comment by timmo001** [comment]
Sorry should have set this to WIP. Still adding and changing, just pushing as I go :smiley:

**Comment by timmo001** [comment]
Here's the prompt in action:

image

image
image
image

**Comment by timmo001** [comment]
OK, let me know what you think

File: src/dialogs/generic/dialog-box.ts [hunk 5]
```diff
@@ -14,61 +14,94 @@ import "../../components/dialog/ha-paper-dialog";
...
       return html``;
     }
 
+    const confirmPrompt = this._params.confirmation || this._params.prompt;
+
     return html`
       <ha-paper-dialog
         with-backdrop
         opened
-        modal
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]: It should be a modal?

File: src/dialogs/generic/dialog-box.ts [hunk 6]
```diff
@@ -14,61 +14,94 @@ import "../../components/dialog/ha-paper-dialog";
...
         @opened-changed="${this._openedChanged}"
       >
         <h2>
           ${this._params.title
             ? this._params.title
-            : this.hass.localize("ui.dialogs.confirmation.title")}
+            : confirmPrompt && this.hass.localize("ui.dialogs.generic.title")}
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]: `Are you sure?` is not a good title for a prompt I think?

Comment (bramkragten) [reply src/dialogs/generic/dialog-box.ts]:
```suggestion
            : this._params.confirmation && this.hass.localize("ui.dialogs.generic.title")}
```

File: src/dialogs/generic/dialog-box.ts [hunk 7]
```diff
@@ -14,61 +14,94 @@ import "../../components/dialog/ha-paper-dialog";
...
+                <p>${this._params.text}</p>
+              `}
         </paper-dialog-scrollable>
         <div class="paper-dialog-buttons">
-          <mwc-button @click="${this._dismiss}">
-            ${this._params.cancelBtnText
-              ? this._params.cancelBtnText
-              : this.hass.localize("ui.dialogs.confirmation.cancel")}
-          </mwc-button>
+          ${confirmPrompt &&
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]:
```
          ${confirmPrompt ? html`...` : ""}
```

File: src/dialogs/generic/dialog-box.ts [hunk 8]
```diff
@@ -14,61 +14,94 @@ import "../../components/dialog/ha-paper-dialog";
...
+                  @value-changed=${this._textChanged}
+                  label=${this._params.inputLabel
+                    ? this._params.inputLabel
+                    : ""}
+                  type=${this._params.inputType
+                    ? this._params.inputType
+                    : "text"}
+                ></paper-input>
+              `
+            : this._params.text &&
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]:
```suggestion
            : this._params.text ? html`...` : ""
```

File: src/dialogs/generic/show-dialog-box.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
+import { fireEvent } from "../../common/dom/fire_event";
+
+export interface AlertDialogParams {
+  title?: string;
+  text?: string;
+  dismissText?: string;
+  confirm?: (out?: string) => void;
```
Comment (bramkragten) [inline src/dialogs/generic/show-dialog-box.ts]: an Alert doesn't have an `out` neither does an confirmation.

File: src/dialogs/generic/show-dialog-box.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  confirmText?: string;
+  cancel?: () => void;
+}
+
+export interface PromptDialogParams extends ConfirmationDialogParams {
+  inputType?: string;
+  inputLabel?: string;
+}
+
+export interface GenericDialogParams
```
Comment (bramkragten) [inline src/dialogs/generic/show-dialog-box.ts]: type GenericDialogParams = PromptDialogParams | ConfirmationDialogParams | AlertDialogParams;

File: src/dialogs/generic/show-dialog-box.ts [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+export const loadGenericDialog = () =>
+  import(/* webpackChunkName: "confirmation" */ "./dialog-box");
+
+export const showAlertDialog = (
+  element: HTMLElement,
+  dialogParams: AlertDialogParams
+): void => {
+  fireEvent(element, "show-dialog", {
```
Comment (bramkragten) [inline src/dialogs/generic/show-dialog-box.ts]: create one functions that fires the event, that takes an argument for the config.
```typescript
showDialog(dialogParams) {
  fireEvent(element, "show-dialog", {
    dialogTag: "dialog-box",
    dialogImport: loadGenericDialog,
    dialogParams,
  });
}
```

File: src/dialogs/generic/dialog-box.ts [hunk 12]
```diff
@@ -14,61 +14,94 @@ import "../../components/dialog/ha-paper-dialog";
...
           ${this._params.title
             ? this._params.title
-            : this.hass.localize("ui.dialogs.confirmation.title")}
+            : confirmPrompt && this.hass.localize("ui.dialogs.generic.title")}
         </h2>
         <paper-dialog-scrollable>
-          <p>${this._params.text}</p>
+          ${this._params.prompt
+            ? html`
+                <paper-input
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]: Add autofocus to the input

**Comment by bramkragten** [comment]
@timmo001 Do you have time to finish this PR?

**Comment by Misiu** [comment]
Any updates?

**Comment by timmo001** [comment]
OK rebase seems to have worked. I'll have a look now and fix anything that pops up as well as address the above

**Comment by timmo001** [comment]
OK, so I've updated to the latest and everything seems to work as it should.

Let me know if there's anything else :+1:

**Comment by bramkragten** [comment]
Sorry for the late response, I was under the impression this was not done yet for review.

File: src/dialogs/generic/show-dialog-box.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+  text?: string;
+  title?: string;
+  cancel?: () => void;
+  confirm?: (out?: string) => void;
+}
+
+export const loadGenericDialog = () =>
+  import(/* webpackChunkName: "confirmation" */ "./dialog-box");
+
+export const showDialog = (
```
Comment (bramkragten) [inline src/dialogs/generic/show-dialog-box.ts]: Create 3 helper functions, for each type of dialog so, `showConfirmationDialog` that sets `confirmation` to true, `showAlertDialog` and `showPromptDialog`.

All with their own params type, so I can not set an input type on a confirm dialog, etc.

File: src/dialogs/generic/dialog-box.ts [hunk 14]
```diff
@@ -42,33 +48,64 @@ class DialogConfirmation extends LitElement {
         <h2>
           ${this._params.title
             ? this._params.title
-            : this.hass.localize("ui.dialogs.confirmation.title")}
+            : this._params.confirmation &&
+              this.hass.localize("ui.dialogs.generic.title")}
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]:
```suggestion
              this.hass.localize("ui.dialogs.generic.default_confirmation_title")}
```

File: src/dialogs/generic/dialog-box.ts [hunk 15]
```diff
@@ -42,33 +48,64 @@ class DialogConfirmation extends LitElement {
...
         </h2>
         <paper-dialog-scrollable>
-          <p>${this._params.text}</p>
+          ${this._params.prompt
+            ? html`
+                <paper-input
+                  autofocus
+                  .value=${this._text}
+                  @value-changed=${this._textChanged}
+                  label=${this._params.inputLabel
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]:
```suggestion
                  .label=${this._params.inputLabel
```

File: src/dialogs/generic/dialog-box.ts [hunk 16]
```diff
@@ -42,33 +48,64 @@ class DialogConfirmation extends LitElement {
...
+                    ? this._params.inputLabel
+                    : ""}
+                  type=${this._params.inputType
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]:
```suggestion
                  .type=${this._params.inputType
```

File: src/dialogs/generic/dialog-box.ts [hunk 17]
```diff
@@ -14,24 +14,30 @@ import "../../components/dialog/ha-paper-dialog";
...
   @property() public hass!: HomeAssistant;
-  @property() private _params?: ConfirmationDialogParams;
+  @property() private _params?: DialogParams;
+  @property() private _text?: string;
 
-  public async showDialog(params: ConfirmationDialogParams): Promise<void> {
+  public async showDialog(params: DialogParams): Promise<void> {
     this._params = params;
+    if (params.prompt) {
+      this._text = params.text;
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]: This should be `params.default_value`

File: src/dialogs/generic/dialog-box.ts [hunk 18]
```diff
@@ -42,33 +48,64 @@ class DialogConfirmation extends LitElement {
...
+                    ? this._params.inputLabel
+                    : ""}
+                  type=${this._params.inputType
+                    ? this._params.inputType
+                    : "text"}
+                ></paper-input>
+              `
+            : this._params.text
+            ? html`
+                <p>${this._params.text}</p>
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]: You can display text when it is a prompt?

File: src/dialogs/generic/dialog-box.ts [hunk 19]
```diff
@@ -42,33 +48,64 @@ class DialogConfirmation extends LitElement {
...
+            : this._params.confirmation &&
+              this.hass.localize("ui.dialogs.generic.title")}
         </h2>
         <paper-dialog-scrollable>
-          <p>${this._params.text}</p>
+          ${this._params.prompt
+            ? html`
+                <paper-input
+                  autofocus
+                  .value=${this._text}
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]:
```suggestion
                  .value=${this._value}
```

File: src/dialogs/generic/dialog-box.ts (hunk 17 above)
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]:
```suggestion
      this._value = params.default_value;
```

**Review by bramkragten — APPROVED** [review]
OK to merge after comments are addressed.

**Comment by timmo001** [comment]
Commited suggestions. I'll take a proper look when I get home and address the rest later as I'm remote at the moment

File: src/dialogs/generic/dialog-box.ts (hunk 18 above)
Comment (timmo001) [reply src/dialogs/generic/dialog-box.ts]: I'll move this above since a prompt might want an explanation above the input

**Comment by timmo001** [comment]
OK, I think that's everything

File: src/dialogs/generic/dialog-box.ts [hunk 20]
```diff
@@ -14,24 +14,30 @@ import "../../components/dialog/ha-paper-dialog";
...
   @property() public hass!: HomeAssistant;
-  @property() private _params?: ConfirmationDialogParams;
+  @property() private _params?: DialogParams;
+  @property() private _value?: string;
 
-  public async showDialog(params: ConfirmationDialogParams): Promise<void> {
+  public async showDialog(params: DialogParams): Promise<void> {
     this._params = params;
+    if (params.prompt) {
+      this._value = params.default_value;
```
Comment (bramkragten) [inline src/dialogs/generic/dialog-box.ts]:
```suggestion
      this._value = params.defaultValue;
```

Comment (bramkragten) [reply src/dialogs/generic/dialog-box.ts]: Sorry, I know it was me 😇

File: src/dialogs/generic/show-dialog-box.ts [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+
+interface ConfirmationDialogParams extends AlertDialogParams {
+  dismissText?: string;
+  cancel?: () => void;
+}
+
+interface PromptDialogParams extends AlertDialogParams {
+  inputLabel?: string;
+  inputType?: string;
+  default_value?: string;
```
Comment (bramkragten) [inline src/dialogs/generic/show-dialog-box.ts]:
```suggestion
  defaultValue?: string;
```
Sorry, I know it was me 😇

File: src/panels/config/users/ha-user-editor.ts [hunk 22]
```diff
@@ -153,14 +153,24 @@ class HaUserEditor extends LitElement {
...
+        )} ${err.message}`,
+      });
     }
   }
 
+  private async _handlePromptRenameUser(ev): Promise<void> {
+    ev.currentTarget.blur();
+    showPromptDialog(this, {
+      title: this.hass!.localize("ui.panel.config.users.editor.enter_new_name"),
+      text: this.user!.name,
```
Comment (bramkragten) [inline src/panels/config/users/ha-user-editor.ts]:
```suggestion
      defaultValue: this.user!.name,
```

**Pull request merged**